lua/lint/parser.lua

```diff
@@ -22,53 +21,85 @@ function M.from_errorformat(efm, skeleton)
           },
           message = item.text,
         }
-        table.insert(
-          result,
-          vim.tbl_extend('keep', diagnostic, skeleton and skeleton or defaults)
-        )
+        table.insert(result, vim.tbl_extend('keep', diagnostic, skeleton and skeleton or defaults))
       end
     end
     return result
   end
 end
 
+--- Parse a linter's output using a regex pattern
+-- @param pattern The regex pattern
+-- @param groups The groups defined by the pattern: {"line", "message", "start_col", ["end_col"], ["code"], ["code_desc"], ["file"], ["severity"]}
+-- @param severity_map An optional table mapping the severity values to their codes
+-- @param defaults An optional table of diagnostic default values
+function M.from_pattern(pattern, groups, severity_map, defaults)
+  local group_handler = {
+    code = function(entries)
+      return entries['code']
+    end,
 
--- Return a parse function that uses a pattern to parse the output.
---
--- The first argument is a lua pattern.
--- The pattern must match 4 groups in order: line number, offset, code and message
---
--- The second argument is a skeleton, used to create each diagnostic.
--- It should contain default values - for example the `source` and `severity`.
---
--- The output of the linter must have 1 entry per line
-function M.from_pattern(pattern, diagnostic_skeleton)
-  return function(output)
-    local result = vim.fn.split(output, "\n")
+    codeDescription = function(entries)
+      return entries['code_desc']
+    end,
+
+    message = function(entries)
+      return entries['message']
+    end,
+
+    range = function(entries)
+      local line = tonumber(entries['line'])
+      local start_col = tonumber(entries['start_col'])
+      local end_col = entries['end_col'] and tonumber(entries['end_col']) or start_col
+      return {
+        ['start'] = { line = line - 1, character = start_col - 1 },
+        ['end'] = { line = line - 1, character = end_col },
+      }
+    end,
+
+    severity = function(entries)
+      return severity_map[entries['severity']] or defaults['severity'] or severity_map['error']
+    end,
+  }
+
+  severity_map = severity_map
+    or {
+      ['error'] = vim.lsp.protocol.DiagnosticSeverity.Error,
+      ['warning'] = vim.lsp.protocol.DiagnosticSeverity.Warning,
+      ['information'] = vim.lsp.protocol.DiagnosticSeverity.Information,
+      ['hint'] = vim.lsp.protocol.DiagnosticSeverity.Hint,
+    }
+  defaults = defaults or {}
+  return function(output, bufnr)
     local diagnostics = {}
+    local buffer_path = vim.api.nvim_buf_get_name(bufnr)
+
+    for _, line in ipairs(vim.fn.split(output, '\n')) do
+      local results = { line:gmatch(pattern)() }
```

Comment:
Out of curiosity, is there a reason to use `line:gmatch(pattern)()` rather than just `line:match(pattern)`? They do seem to be different, as with the former (current) implementation I am not getting any results from my [nix](https://github.com/mfussenegger/nvim-lint/pull/44) linter, but with the latter implementation it does work.